File: packages/lhc/src/thread-view/internal/protected-boundary.ts
/**
 * Read-only protected visibility-boundary preview (LIM-67).
 *
 * Accounts every protected tool_result at full size first, then considers only
 * older unprotected tool_result rows after the effective compact/boundary start.
 * Never moves the boundary backward. Boundary is strictly before the earliest
 * protected result event. Calls, assistant text, and reasoning are never targets.
 */

import type { DatabaseSync } from "node:sqlite";
import { estimateTokens } from "../../shared-tech/token-counting/index.js";
import { readBoundaryPosition } from "./boundary.js";
import { ABBREVIATION_LIMIT, deterministicTruncation } from "./render.js";
import { readViewSnapshot } from "./snapshot.js";

export type ProtectedBoundaryPreview = {
  previousBoundary: number;
  proposedBoundary: number;
  compactPoint: number;
  effectiveStart: number;
  /** Earliest protected tool_result source_event_order; null when none. */
  earliestProtectedResultOrder: number | null;
  /** Max legal boundary is earliestProtectedResultOrder - 1 (or previous when none). */
  maxLegalBoundary: number;
  protectedToolCallIds: string[];
  fullProtectedTokenEstimate: number;
  eligibleUnprotectedTokenEstimate: number;
  prunedUnprotectedTokenEstimate: number;
  remainingUnprotectedFullTokenEstimate: number;
  /** Estimated zone tokens after proposed boundary (protected full + remaining full unprotected). */
  zoneTokensAfterEstimate: number;
  /** True when even maximal eligible pruning cannot reduce zone under target (if any). */
  maximalPruneInsufficient: boolean;
  noOp: boolean;
  /** Human-readable reasons (empty when ok). */
  notes: string[];
};

type ToolResultRow = {
  sourceEventOrder: number;
  toolCallId: string;
  tokenEstimate: number;
  content: string;
};

function readLiveToolResultsAfter(db: DatabaseSync, effectiveStart: number): ToolResultRow[] {
  const rows = db
    .prepare(
      `SELECT m.source_event_order, m.token_estimate, b.content
       FROM message m
       JOIN message_block b ON b.message_id = m.message_id AND b.block_type = 'tool_result'
       WHERE m.kind = 'tool_result' AND m.deleted_at IS NULL AND m.source_event_order > ?
       ORDER BY m.source_event_order ASC`,
    )
    .all(effectiveStart) as unknown as Array<{
    source_event_order: number | bigint;
    token_estimate: number | bigint;
    content: string;
  }>;
  return rows.map((row) => {
    let toolCallId = "";
    let content = "";
    try {
      const parsed = JSON.parse(row.content) as { toolCallId?: unknown; content?: unknown };
      if (typeof parsed.toolCallId === "string") toolCallId = parsed.toolCallId;
      if (typeof parsed.content === "string") content = parsed.content;
    } catch {
      // leave empty; caller may still count token_estimate
    }
    return {
      sourceEventOrder: Number(row.source_event_order),
      toolCallId,
      tokenEstimate: Number(row.token_estimate),
      content,
    };
  });
}

function abridgedTokenEstimate(content: string): number {
  return estimateTokens(deterministicTruncation(content));
}

/**
 * Compute a monotonic proposed visibility boundary that:
 * - starts at max(previousBoundary, compactPoint)
 * - never moves backward
 * - remains strictly before the earliest protected result event
 * - only advances across older unprotected tool_result rows
 *
 * When `targetZoneTokens` is provided, advances just far enough that the
 * remaining full zone (full protected + full unprotected ahead of boundary)
 * is at or under the target, preferring the smallest advance that meets it.
 * When omitted, computes the maximal legal advance (for insufficiency proofs).
 */
export function previewProtectedVisibilityBoundary(
  db: DatabaseSync,
  protectedToolCallIds: readonly string[],
  opts: { targetZoneTokens?: number; compactPointOverride?: number } = {},
): ProtectedBoundaryPreview {
  const protectedSet = new Set(protectedToolCallIds.filter((id) => typeof id === "string" && id.length > 0));
  const sortedProtected = [...protectedSet].sort();
  const snapshot = readViewSnapshot(db);
  const compactPoint = opts.compactPointOverride ?? snapshot?.compactPoint ?? 0;
  const previousBoundary = readBoundaryPosition(db);
  const effectiveStart = Math.max(previousBoundary, compactPoint);
  const notes: string[] = [];

  const rows = readLiveToolResultsAfter(db, effectiveStart);
  const protectedRows = rows.filter((r) => protectedSet.has(r.toolCallId));
  const unprotectedRows = rows.filter((r) => !protectedSet.has(r.toolCallId));

  const fullProtectedTokenEstimate = protectedRows.reduce((s, r) => s + r.tokenEstimate, 0);
  const eligibleUnprotectedTokenEstimate = unprotectedRows.reduce((s, r) => s + r.tokenEstimate, 0);

  let earliestProtectedResultOrder: number | null = null;
  for (const r of protectedRows) {
    if (earliestProtectedResultOrder === null || r.sourceEventOrder < earliestProtectedResultOrder) {
      earliestProtectedResultOrder = r.sourceEventOrder;
    }
  }

  // Strictly before earliest protected result; if none, may advance through all eligible.
  const maxLegalBoundary =
    earliestProtectedResultOrder === null ? Number.MAX_SAFE_INTEGER : earliestProtectedResultOrder - 1;

  if (earliestProtectedResultOrder !== null && earliestProtectedResultOrder <= effectiveStart) {
    notes.push("earliest protected result is at or behind effective start; no eligible older unprotected rows");
  }

  // Eligible unprotected rows that can be crossed (result order <= maxLegalBoundary).
  const eligible = unprotectedRows.filter((r) => r.sourceEventOrder <= maxLegalBoundary);

  // Maximal advance: last eligible unprotected result order (or previous if none).
  let maximalBoundary = previousBoundary;
  for (const r of eligible) {
    if (r.sourceEventOrder > maximalBoundary && r.sourceEventOrder <= maxLegalBoundary) {
      maximalBoundary = r.sourceEventOrder;
    }
  }
  if (maximalBoundary < previousBoundary) maximalBoundary = previousBoundary;
  if (maximalBoundary > maxLegalBoundary && maxLegalBoundary >= previousBoundary) {
    maximalBoundary = Math.max(previousBoundary, Math.min(maximalBoundary, maxLegalBoundary));
  }
  // Boundary must stay strictly before protected results.
  if (earliestProtectedResultOrder !== null && maximalBoundary >= earliestProtectedResultOrder) {
    maximalBoundary = Math.max(previousBoundary, earliestProtectedResultOrder - 1);
  }

  function zoneFullAfter(boundary: number): number {
    // Full tokens for results with source_event_order > max(boundary, compactPoint)
    const start = Math.max(boundary, compactPoint);
    let sum = 0;
    for (const r of rows) {
      if (r.sourceEventOrder > start) sum += r.tokenEstimate;
    }
    return sum;
  }

  function prunedSavingsIfBoundary(boundary: number): number {
    // Results at-or-behind boundary and after compact point render abridged.
    let savings = 0;
    for (const r of unprotectedRows) {
      if (r.sourceEventOrder > compactPoint && r.sourceEventOrder <= boundary) {
        const full = r.tokenEstimate;
        const short = abridgedTokenEstimate(r.content);
        savings += Math.max(0, full - short);
      }
    }
    // Protected must never be abridged — if boundary would abridge them, illegal.
    return savings;
  }

  let proposedBoundary = previousBoundary;
  const target = opts.targetZoneTokens;

  if (target === undefined) {
    proposedBoundary = maximalBoundary;
  } else if (zoneFullAfter(previousBoundary) <= target) {
    proposedBoundary = previousBoundary;
  } else {
    // Walk eligible rows oldest-first, advancing boundary to each unprotected result
    // until zone is under target or maximal legal boundary is reached.
    proposedBoundary = previousBoundary;
    for (const r of eligible) {
      if (r.sourceEventOrder <= proposedBoundary) continue;
      if (r.sourceEventOrder > maxLegalBoundary) break;
      const candidate = r.sourceEventOrder;
      if (earliestProtectedResultOrder !== null && candidate >= earliestProtectedResultOrder) break;
      proposedBoundary = candidate;
      if (zoneFullAfter(proposedBoundary) <= target) break;
    }
  }

  // Never move backward; clamp to legal.
  if (proposedBoundary < previousBoundary) proposedBoundary = previousBoundary;
  if (earliestProtectedResultOrder !== null && proposedBoundary >= earliestProtectedResultOrder) {
    proposedBoundary = Math.max(previousBoundary, earliestProtectedResultOrder - 1);
    notes.push("clamped proposed boundary strictly before earliest protected result");
  }

  // Pruned estimate: eligible unprotected at-or-behind proposed boundary.
  let prunedUnprotectedTokenEstimate = 0;
  let remainingUnprotectedFullTokenEstimate = 0;
  for (const r of unprotectedRows) {
    if (r.sourceEventOrder > Math.max(proposedBoundary, compactPoint)) {
      remainingUnprotectedFullTokenEstimate += r.tokenEstimate;
    } else if (r.sourceEventOrder > compactPoint && r.sourceEventOrder <= proposedBoundary) {
      prunedUnprotectedTokenEstimate += r.tokenEstimate;
    }
  }

  const zoneTokensAfterEstimate = zoneFullAfter(proposedBoundary);
  const maximalZone = zoneFullAfter(maximalBoundary);
  const maximalPruneInsufficient =
    target !== undefined ? maximalZone > target : fullProtectedTokenEstimate > 0 && eligible.length === 0;

  if (earliestProtectedResultOrder !== null && proposedBoundary >= earliestProtectedResultOrder) {
    // Should be impossible after clamp; hard guard.
    proposedBoundary = Math.max(previousBoundary, earliestProtectedResultOrder - 1);
  }

  const noOp = proposedBoundary === previousBoundary;
  void prunedSavingsIfBoundary;
  void ABBREVIATION_LIMIT;

  return {
    previousBoundary,
    proposedBoundary,
    compactPoint,
    effectiveStart,
    earliestProtectedResultOrder,
    maxLegalBoundary: earliestProtectedResultOrder === null ? maximalBoundary : earliestProtectedResultOrder - 1,
    protectedToolCallIds: sortedProtected,
    fullProtectedTokenEstimate,
    eligibleUnprotectedTokenEstimate,
    prunedUnprotectedTokenEstimate,
    remainingUnprotectedFullTokenEstimate,
    zoneTokensAfterEstimate,
    maximalPruneInsufficient,
    noOp,
    notes,
  };
}

File: packages/lhc-rs/src/thread_view/internal/protected_boundary.rs
//! Read-only protected visibility-boundary preview (LIM-67).
//!
//! Ported from `packages/lhc/src/thread-view/internal/protected-boundary.ts`.
//! Accounts every protected tool_result at full size first, then considers only
//! older unprotected tool_result rows after the effective compact/boundary start.
//! Never moves the boundary backward. Boundary is strictly before the earliest
//! protected result event. Calls, assistant text, and reasoning are never targets.

use std::collections::BTreeSet;

use serde::{Deserialize, Serialize};

use crate::shared_tech::storage::{Db, SqlParam};
use crate::thread_view::internal::boundary::read_boundary_position;
use crate::thread_view::internal::snapshot::read_view_snapshot;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProtectedBoundaryPreview {
    pub previous_boundary: i64,
    pub proposed_boundary: i64,
    pub compact_point: i64,
    pub effective_start: i64,
    /// Earliest protected tool_result source_event_order; None when none.
    pub earliest_protected_result_order: Option<i64>,
    /// Max legal boundary is earliest protected order - 1 (or maximal when none).
    pub max_legal_boundary: i64,
    pub protected_tool_call_ids: Vec<String>,
    pub full_protected_token_estimate: i64,
    pub eligible_unprotected_token_estimate: i64,
    pub pruned_unprotected_token_estimate: i64,
    pub remaining_unprotected_full_token_estimate: i64,
    /// Estimated zone tokens after proposed boundary (protected full + remaining full unprotected).
    pub zone_tokens_after_estimate: i64,
    /// True when even maximal eligible pruning cannot reduce zone under target (if any).
    pub maximal_prune_insufficient: bool,
    pub no_op: bool,
    /// Human-readable reasons (empty when ok).
    pub notes: Vec<String>,
}

#[derive(Debug, Clone)]
struct ToolResultRow {
    source_event_order: i64,
    tool_call_id: String,
    token_estimate: i64,
}

#[derive(Debug, Clone, Default)]
pub struct ProtectedBoundaryOpts {
    pub target_zone_tokens: Option<i64>,
    pub compact_point_override: Option<i64>,
}

fn read_live_tool_results_after(db: &Db, effective_start: i64) -> Vec<ToolResultRow> {
    let rows = db
        .prepare(
            "SELECT m.source_event_order, m.token_estimate, b.content
             FROM message m
             JOIN message_block b ON b.message_id = m.message_id AND b.block_type = 'tool_result'
             WHERE m.kind = 'tool_result' AND m.deleted_at IS NULL AND m.source_event_order > ?
             ORDER BY m.source_event_order ASC",
        )
        .all(&[SqlParam::from(effective_start)]);
    rows.iter()
        .map(|row| {
            let mut tool_call_id = String::new();
            if let Some(raw) = row.get("content").and_then(|v| v.as_str())
                && let Ok(parsed) = serde_json::from_str::<serde_json::Value>(raw)
                && let Some(id) = parsed.get("toolCallId").and_then(|v| v.as_str())
            {
                tool_call_id = id.to_string();
            }
            ToolResultRow {
                source_event_order: row
                    .get("source_event_order")
                    .and_then(|v| v.as_i64())
                    .unwrap_or(0),
                tool_call_id,
                token_estimate: row
                    .get("token_estimate")
                    .and_then(|v| v.as_i64())
                    .unwrap_or(0),
            }
        })
        .collect()
}

/// Compute a monotonic proposed visibility boundary that:
/// - starts at max(previousBoundary, compactPoint)
/// - never moves backward
/// - remains strictly before the earliest protected result event
/// - only advances across older unprotected tool_result rows
///
/// When `target_zone_tokens` is provided, advances just far enough that the
/// remaining full zone is at or under the target, preferring the smallest
/// advance that meets it. When omitted, computes the maximal legal advance.
pub fn preview_protected_visibility_boundary(
    db: &Db,
    protected_tool_call_ids: &[String],
    opts: &ProtectedBoundaryOpts,
) -> ProtectedBoundaryPreview {
    let protected_set: BTreeSet<String> = protected_tool_call_ids
        .iter()
        .filter(|id| !id.is_empty())
        .cloned()
        .collect();
    let mut sorted_protected: Vec<String> = protected_set.iter().cloned().collect();
    sorted_protected.sort_by(|a, b| crate::shared_tech::compact_continuation::js_string_cmp(a, b));
    let snapshot = read_view_snapshot(db);
    let compact_point = opts
        .compact_point_override
        .unwrap_or_else(|| snapshot.map(|s| s.compact_point).unwrap_or(0));
    let previous_boundary = read_boundary_position(db);
    let effective_start = previous_boundary.max(compact_point);
    let mut notes: Vec<String> = Vec::new();

    let rows = read_live_tool_results_after(db, effective_start);
    let protected_rows: Vec<&ToolResultRow> = rows
        .iter()
        .filter(|r| protected_set.contains(&r.tool_call_id))
        .collect();
    let unprotected_rows: Vec<&ToolResultRow> = rows
        .iter()
        .filter(|r| !protected_set.contains(&r.tool_call_id))
        .collect();

    let full_protected_token_estimate: i64 = protected_rows.iter().map(|r| r.token_estimate).sum();
    let eligible_unprotected_token_estimate: i64 =
        unprotected_rows.iter().map(|r| r.token_estimate).sum();

    let earliest_protected_result_order: Option<i64> =
        protected_rows.iter().map(|r| r.source_event_order).min();

    // Strictly before earliest protected result; if none, may advance through all eligible.
    let max_legal_boundary = match earliest_protected_result_order {
        None => i64::MAX,
        Some(order) => order - 1,
    };

    if let Some(order) = earliest_protected_result_order
        && order <= effective_start
    {
        notes.push(
            "earliest protected result is at or behind effective start; no eligible older unprotected rows"
                .to_string(),
        );
    }

    // Eligible unprotected rows that can be crossed.
    let eligible: Vec<&ToolResultRow> = unprotected_rows
        .iter()
        .filter(|r| r.source_event_order <= max_legal_boundary)
        .copied()
        .collect();

    // Maximal advance: last eligible unprotected result order (or previous if none).
    let mut maximal_boundary = previous_boundary;
    for r in &eligible {
        if r.source_event_order > maximal_boundary && r.source_event_order <= max_legal_boundary {
            maximal_boundary = r.source_event_order;
        }
    }
    if maximal_boundary < previous_boundary {
        maximal_boundary = previous_boundary;
    }
    if maximal_boundary > max_legal_boundary && max_legal_boundary >= previous_boundary {
        maximal_boundary = previous_boundary.max(maximal_boundary.min(max_legal_boundary));
    }
    if let Some(order) = earliest_protected_result_order
        && maximal_boundary >= order
    {
        maximal_boundary = previous_boundary.max(order - 1);
    }

    let zone_full_after = |boundary: i64| -> i64 {
        let start = boundary.max(compact_point);
        rows.iter()
            .filter(|r| r.source_event_order > start)
            .map(|r| r.token_estimate)
            .sum()
    };

    let mut proposed_boundary = previous_boundary;
    let target = opts.target_zone_tokens;

    match target {
        None => {
            proposed_boundary = maximal_boundary;
        }
        Some(target) if zone_full_after(previous_boundary) <= target => {
            proposed_boundary = previous_boundary;
        }
        Some(target) => {
            // Walk eligible rows oldest-first until zone is under target or maximal
            // legal boundary is reached.
            for r in &eligible {
                if r.source_event_order <= proposed_boundary {
                    continue;
                }
                if r.source_event_order > max_legal_boundary {
                    break;
                }
                let candidate = r.source_event_order;
                if let Some(order) = earliest_protected_result_order
                    && candidate >= order
                {
                    break;
                }
                proposed_boundary = candidate;
                if zone_full_after(proposed_boundary) <= target {
                    break;
                }
            }
        }
    }

    // Never move backward; clamp to legal.
    if proposed_boundary < previous_boundary {
        proposed_boundary = previous_boundary;
    }
    if let Some(order) = earliest_protected_result_order
        && proposed_boundary >= order
    {
        proposed_boundary = previous_boundary.max(order - 1);
        notes.push(
            "clamped proposed boundary strictly before earliest protected result".to_string(),
        );
    }

    // Pruned estimate: eligible unprotected at-or-behind proposed boundary.
    let mut pruned_unprotected_token_estimate: i64 = 0;
    let mut remaining_unprotected_full_token_estimate: i64 = 0;
    for r in &unprotected_rows {
        if r.source_event_order > proposed_boundary.max(compact_point) {
            remaining_unprotected_full_token_estimate += r.token_estimate;
        } else if r.source_event_order > compact_point && r.source_event_order <= proposed_boundary
        {
            pruned_unprotected_token_estimate += r.token_estimate;
        }
    }

    let zone_tokens_after_estimate = zone_full_after(proposed_boundary);
    let maximal_zone = zone_full_after(maximal_boundary);
    let maximal_prune_insufficient = match target {
        Some(target) => maximal_zone > target,
        None => full_protected_token_estimate > 0 && eligible.is_empty(),
    };

    if let Some(order) = earliest_protected_result_order
        && proposed_boundary >= order
    {
        // Should be impossible after clamp; hard guard.
        proposed_boundary = previous_boundary.max(order - 1);
    }

    let no_op = proposed_boundary == previous_boundary;

    ProtectedBoundaryPreview {
        previous_boundary,
        proposed_boundary,
        compact_point,
        effective_start,
        earliest_protected_result_order,
        max_legal_boundary: match earliest_protected_result_order {
            None => maximal_boundary,
            Some(order) => order - 1,
        },
        protected_tool_call_ids: sorted_protected,
        full_protected_token_estimate,
        eligible_unprotected_token_estimate,
        pruned_unprotected_token_estimate,
        remaining_unprotected_full_token_estimate,
        zone_tokens_after_estimate,
        maximal_prune_insufficient,
        no_op,
        notes,
    }
}
